core/src/main/java/ai/timefold/solver/core/impl/domain/variable/declarative/VariableReferenceGraph.java

```diff
@@ -0,0 +1,305 @@
+        variableReferenceToAfterProcessor.computeIfAbsent(parentVariableId, k -> new ArrayList<>())
+                .add(consumer);
+    }
+
+    public void createGraph(IntFunction<TopologicalOrderGraph> graphCreator) {
+        counts = new int[instanceList.size()][instanceList.size()];
+        graph = graphCreator.apply(instanceList.size());
+        graph.withNodeData(instanceList);
+        changed = new BitSet(instanceList.size());
+
+        graph.startBatchChange();
+        var visited = Collections.newSetFromMap(new IdentityHashMap<>());
+        for (var instance : instanceList) {
+            if (visited.add(instance.entity())) {
+                for (var variableId : variableReferenceToAfterProcessor.keySet()) {
+                    if (variableId.entityClass().isInstance(instance.entity())) {
+                        afterVariableChanged(variableId, instance.entity());
+                    }
+                }
+            }
+            entityToVariableReferenceMap.computeIfAbsent(instance.entity(), ignored -> new ArrayList<>())
+                    .add(instance);
+        }
+        for (var fixedEdgeEntry : fixedEdges.entrySet()) {
+            for (var toEdge : fixedEdgeEntry.getValue()) {
+                addEdge(fixedEdgeEntry.getKey(), toEdge);
+            }
+        }
+    }
+
+    public EntityVariableOrFactReference<?> lookup(VariableId variableReference, Object entity) {
+        return variableReferenceToInstanceMap.getOrDefault(variableReference, Collections.emptyMap()).get(entity);
+    }
+
+    public void addFixedEdge(EntityVariableOrFactReference<Solution_> from, EntityVariableOrFactReference<Solution_> to) {
+        if (from.id() == to.id()) {
+            return;
+        }
+        fixedEdges.computeIfAbsent(from, k -> new ArrayList<>()).add(to);
+    }
+
+    public void addEdge(EntityVariableOrFactReference<Solution_> from, EntityVariableOrFactReference<Solution_> to) {
+        if (from.id() == to.id()) {
+            return;
+        }
+        if (changed.isEmpty()) {
+            graph.startBatchChange();
+        }
+        var oldCount = counts[from.id()][to.id()]++;
+        if (oldCount == 0) {
+            graph.addEdge(from.id(), to.id());
+        }
+
+        markChanged(to);
+    }
+
+    public void removeEdge(EntityVariableOrFactReference<Solution_> from, EntityVariableOrFactReference<Solution_> to) {
+        if (from.id() == to.id()) {
+            return;
+        }
+        if (changed.isEmpty()) {
+            graph.startBatchChange();
+        }
+        var newCount = --counts[from.id()][to.id()];
+        if (newCount == 0) {
+            graph.removeEdge(from.id(), to.id());
+        }
+        markChanged(to);
+    }
+
+    public void markChanged(EntityVariableOrFactReference<?> node) {
+        if (changed.isEmpty()) {
+            graph.startBatchChange();
+        }
+        changed.set(node.id());
+    }
+
+    public void updateChanged() {
```

Comment:
This method is very long and also - not coincidentally - shows up as a dominator in the flamegraph. Please break this method down into smaller methods, so that we can better understand the logic, and so that the flame graph can actually meaningfully show how much each part contributes to the result.